### PR TITLE
Feature/Refactor UIPOptions

### PR DESCRIPTION
**Pull request by Sisha3342**

### CODE REVIEW

File: src/core/base/root.ts [hunk 1]
```diff
@@ -20,22 +20,21 @@ export class UIPRoot extends ESLBaseElement {
    * Attribute for controlling UIP components' theme.
    * Has two values: `uip-light` and `uip-dark`.
    */
```
Comment (yadamskaya) [inline src/core/base/root.ts:22]:
```suggestion
   * Indicates that the UIP components' theme is dark.
   */
```

File: src/core/base/root.ts [hunk 2]
```diff
@@ -20,22 +20,21 @@ export class UIPRoot extends ESLBaseElement {
...
-  /** Attribute for editor's visibility state. */
-  @boolAttr() public editor: boolean;
-
+  @boolAttr() public darkTheme: boolean;
+  /**
+   * Attributes for settings, editor, visibility state.
+   * Has two values: `expanded` and `collapsed`.
+   */
+  @boolAttr() public settingsCollapsed: boolean;
+  @boolAttr() public editorCollapsed: boolean;
```
Comment (yadamskaya) [inline src/core/base/root.ts:29]:
```suggestion
  /**
   * Collapsed settings state marker.
   */
  @boolAttr() public settingsCollapsed: boolean;
    /**
   * Collapsed editor state marker.
   */
  @boolAttr() public editorCollapsed: boolean;
```

File: src/core/base/root.ts [hunk 3]
```diff
@@ -20,22 +20,21 @@ export class UIPRoot extends ESLBaseElement {
...
   /**
    * Attribute for controlling preview's content direction.
    * Has two values: `LTR` and `RTL`.
    */
```
Comment (yadamskaya) [inline src/core/base/root.ts:33]:
```suggestion
   * Indicates that the direction of the preview content is RTL direction.
   */
```

File: src/core/base/root.ts [hunk 4]
```diff
@@ -78,9 +71,8 @@ export class UIPRoot extends ESLBaseElement {
     });
   }
 
-  protected _updateStyles(option: string, prev: string, next: string) {
-    this.classList.remove(`${prev}-${option}`);
-    this.classList.add(`${next}-${option}`);
+  protected _updateStyles(option: string, value: string) {
```
Comment (yadamskaya) [inline src/core/base/root.ts:74]:
```suggestion
  private _updateStyles(option: string, value: string) {
```

File: src/core/base/root.ts [hunk 5]
```diff
@@ -78,9 +71,8 @@ export class UIPRoot extends ESLBaseElement {
...
+    value === null ? this.classList.remove(option) : this.classList.add(option);
```
Comment (yadamskaya) [inline src/core/base/root.ts:75]:
```suggestion
    this.classList.toggle(option, value !== null);
```

File: src/plugins/editor/editor.ts [hunk 6]
```diff
@@ -125,25 +120,15 @@ export class UIPEditor extends UIPPlugin {
...
-
-    if (attr === 'editor') {
-      this.toggleEditor();
+    switch (attr) {
+      case 'dark-theme':
+        return this.setEditorConfig({theme: value === null ?
+          UIPEditor.defaultOptions.theme : UIPEditor.themesMapping['uip-dark']});
+      case 'editor-collapsed':
+        return value === null ?
+          this.classList.remove('collapsed') : this.classList.add('collapsed');
```
Comment (yadamskaya) [inline src/plugins/editor/editor.ts:129]:
```suggestion
        return this.classList.toggle('collapsed', value !== null);
```

File: src/plugins/settings/settings.ts [hunk 7]
```diff
@@ -76,13 +75,7 @@ export class UIPSettings extends UIPPlugin {
...
-    if (e.detail.attribute !== 'settings') return;
-    this.toggleSettings();
-  }
-
-  protected toggleSettings(): void {
-    this.root?.settings
-      ? this.classList.remove('collapsed')
-      : this.classList.add('collapsed');
+    if (e.detail.attribute !== 'settings-collapsed') return false;
+    e.detail.value === null ? this.classList.remove('collapsed') : this.classList.add('collapsed');
```
Comment (yadamskaya) [inline src/plugins/settings/settings.ts:79]:
```suggestion
    this.classList.toggle('collapsed', e.detail.value !== null);
```

File: src/core/base/root.ts (hunk 5 above)
Comment (Sisha3342) [reply src/core/base/root.ts:67]: 'force' parameter doesn't work in IE

File: src/plugins/editor/editor.ts (hunk 6 above)
Comment (Sisha3342) [reply src/plugins/editor/editor.ts:129]: 'force' parameter doesn't work in IE

File: src/plugins/settings/settings.ts (hunk 7 above)
Comment (Sisha3342) [reply src/plugins/settings/settings.ts:79]: 'force' parameter doesn't work in IE

File: src/plugins/options/option/theme.less [hunk 8]
```diff
@@ -1,43 +1,51 @@
...
-          background-color: @dark-theme;
-        }
+      & > .uip-plugin-inner {
+        background-color: @dark-theme-section;
+      }
 
-        &-control {
-          &:before {
-            background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 9' xml:space='preserve'%3E%3Cpolyline points='0,0 8,8 16,0' fill='none' stroke='white' stroke-width='2'%3E%3C/polyline%3E%3C/svg%3E");
+      &.uip-header {
```
Comment (nattallius) [inline src/plugins/options/option/theme.less:34]: I still see no reason to make header a plugin in out current implementation. here and in other places at least we can use tag selector or at all not clarify placement in header.

File: src/plugins/options/options.ts [hunk 9]
```diff
@@ -1,119 +1,90 @@
...
+import {OptionConfig, UIPOption} from './option/option';
 
 /**
  * Custom element to provide controls for changing UIP visual appearance.
  * @extends UIPPlugin
  */
 export class UIPOptions extends UIPPlugin {
   static is = 'uip-options';
+  options: Map<string, UIPOption>;
+  optionsConfig: OptionConfig[] = [
```
Comment (nattallius) [inline src/plugins/options/options.ts:13]: minor cosmetic, maybe it makes sense to name it  `UIPOptionConfig`, cause line upper options type is `Map<string, UIPOption>` , for me it was a little bit confusing

File: src/plugins/options/options.ts [hunk 10]
```diff
@@ -1,119 +1,90 @@
...
-    const target = e.target as HTMLInputElement;
-
-    const theme = target.getAttribute('theme');
-    const settings = target.getAttribute('settings-control');
-    const editor = target.getAttribute('editor-control');
-    const dir = target.getAttribute('direction');
+  protected onOptionClick(e: Event) {
+    const option = e.target;
+    if (!(option instanceof UIPOption)) return;
+    this.root?.toggleAttribute(option.attribute, option.active);
```
Comment (nattallius) [inline src/plugins/options/options.ts:83]: If we anyway incidentally have a check of event target here, then i think we can not have multiple event listeners, but do toggling here
```suggestion
    if (!(option instanceof UIPOption)) return;
    option.toggleState();
    this.root?.toggleAttribute(option.attribute, option.active);
```

File: src/plugins/options/options.ts [hunk 11]
```diff
@@ -1,119 +1,90 @@
...
 
-    if (this.root) {
-      if (theme) this.root.theme = theme;
-      if (settings) this.root.settings = target.checked;
-      if (editor) this.root.editor = target.checked;
-      if (dir) this.root.direction = dir;
-    }
+  public static register(tagName?: string) {
+    UIPOption.register();
+    customElements.whenDefined(UIPOption.is).then(() => super.register.call(this, tagName));
```
Comment (nattallius) [inline src/plugins/options/options.ts:88]:
```suggestion
    UIPOption.registered.then(() => super.register.call(this, tagName));
```

**Review by PolinaVolchetskaya — APPROVED** [review]

File: src/core/base/root.ts (hunk 5 above)
Comment (ala-n) [reply src/core/base/root.ts:67]: Guys you have no need to care about IE so much, this one basically can be easily covered by polyfill for IE.

**Review by AlexanderBazukevich — APPROVED** [review]

**Review by yadamskaya — APPROVED** [review]

**Review by PolinaVolchetskaya — APPROVED** [review]

**Pull request merged**